test/e2e: Reduce need to use local kubectl

okartau opened the pull request:
Use POST method instead of running local kunectl.
Use ExecCommandInContainerWithFullOutput() instead of RunKubectl().
PodExec() takes additional framework arg, passed down in call chain.
New variants of two functions in e2e/framework/pods.go:
ExecWithOptions, ExecCommandInContainerWithFullOutput
with added ClientSet, Namespace arguments.
VerifyExecInPodFail uses different error code cast as original
one causes test code Panic if used with new call method.

/kind feature

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Removed dependency on kubectl from several storage E2E tests
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

okartau committed:
test/e2e: Reduce need to use local kubectl

Use POST method instead of running local kubectl.
Use ExecCommandInContainerWithFullOutput() instead of RunKubectl().
PodExec() takes additional framework arg, passed down in call chain.
VerifyExecInPodFail uses different error code cast as original
one causes test code Panic if used with new call method.

BenTheElder commented:
we have conformance tests specifically about using kubectl and that's unlikely to change.

if you want to not use a local kubectl, have you considered using the confromance image and / or sonobuoy?

BenTheElder commented:
I see, I might have suggested one that does NOT import k8s.io/kubernetes ...

On Thu, Oct 17, 2019 at 11:37 AM Patrick Ohly <notifications@github.com>
wrote:

> we have conformance tests specifically about using kubectl and that's
> unlikely to change.
>
> That's fine for tests that are specifically meant to cover Kubernetes and
> it's command line client. But the storage tests should use the Kubernetes
> API directly. That makes them more versatile and easier to reuse.
>
> if you want to not use a local kubectl, have you considered using the
> confromance image and / or sonobuoy?
>
> We are writing our own test suite 😁
>
> See
> https://kubernetes.io/blog/2019/03/22/kubernetes-end-to-end-testing-for-everyone/
> and https://github.com/intel/pmem-csi/tree/devel/test/e2e.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/84042?email_source=notifications&email_token=AAHADK34NCBJQ2Q6DEZ7MWTQPCWGLA5CNFSM4JBYBHRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBRCUYQ#issuecomment-543304290>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHADK7H3EDTVOGT74LNONLQPCWGLANCNFSM4JBYBHRA>
> .
>

pohly commented:
> I see, I might have suggested one that does NOT import k8s.io/kubernetes ...

Another test framework? But then it wouldn't have the same tests readily available, would it?

Anyway, work is underway to move the test/e2e/framework into staging, at which point importing it will be easier.

pohly commented:
/ok-to-test

okartau committed:
utils.go: Changed bashExec to shExec, bash not guaranteed.

okartau committed:
fixtures.go: Removed unused parameter "client" in testVolumeContent()

pohly commented:
/lgtm

I'll let others approve, to ensure that we have consensus that this is a useful change.

dims commented:
/release-note-none
/assign @timothysc

pohly commented on test/e2e/framework/volume/fixtures.go:
```diff
@@ -563,20 +563,20 @@ func testVolumeContent(client clientset.Interface, pod *v1.Pod, fsGroup *int64,
 // Multiple Tests can be specified to mount multiple volumes to a single
 // pod.
-func TestVolumeClient(client clientset.Interface, config TestConfig, fsGroup *int64, fsType string, tests []Test) {
+func TestVolumeClient(f *framework.Framework, client clientset.Interface, config TestConfig, fsGroup *int64, fsType string, tests []Test) {
```
/lgtm cancel

I think this is another case of a redundant `client` parameter. Remove the parameter and use `f.ClientSet` instead in the call to `runVolumeTesterPod`.

okartau replied:
made that change and pushed.
Should we also delete cs arg from test/e2e/storage/flexvolume.go:testFlexVolume, which now becomes unused. By same logic, passing cs is redundant already in current code if we assume it is always set to f.ClientSet, and framework arg is there now already

pohly replied:
> Should we also delete cs arg from test/e2e/storage/flexvolume.go:testFlexVolume, which now becomes unused.

Yes.

> By same logic, passing cs is redundant already in current code if we assume it is always set to f.ClientSet, and framework arg is there now already

Correct.

okartau committed:
Removed 2nd arg "cs" from TestVolumeClient

as we assume it is always set to f.ClientSet and we can
use its value from passed framework argument.

pohly commented:
/lgtm

pohly commented:
/lgtm cancel

timothysc commented:
/assign @alejandrox1

okartau committed:
Remove argument "cs clientset.Interface" from testFlexVolume

cs is not used any more after previous change removed cs arg
from TectVolumeClient. Functions down the call path
get cs value from framework parameter.

pohly reviewed (approved):
/lgtm

pohly commented:
/lgtm

okartau commented:
My recent force-push from 5h ago was done after manual merge-to-HEAD conflict resolve.

cofyc commented on test/e2e/framework/pods.go:
```diff
@@ -313,3 +316,54 @@ func (c *PodClient) PodIsReady(name string) bool {
+// ExecWithOptions executes a command in the specified container,
+// returning stdout, stderr and error. `options` allowed for
+// additional parameters to be passed.
+func (c *PodClient) ExecWithOptions(options ExecOptions, cs clientset.Interface) (string, string, error) {
```
I have two questions:

1)  why we need two new methods on PodClient? These two methods do not access PodClient fields. And we already have [exec methods on Framework](https://github.com/kubernetes/kubernetes/blob/ed5b03831385b0071b127ff6c329ae344b9780f6/test/e2e/framework/exec_util.go#L50), I see the usage of `PodClient()` [here](https://github.com/kubernetes/kubernetes/pull/84042/files#diff-0d21a89b3adc375caf1b557b014e1111R68), It still use the framework ClientSet. 

2) If we need to, I think it's better to have some subroutines and reuse code.

pohly replied:
I might have suggested that to @okartau as one option before he started working on the code. But as you noticed, in the end the resulting methods are not related to `PodClient`, which indeed indicates that they belong elsewhere.

But before we discuss where a better place for them might be, do we actually still need this PR? It was supposed to solve https://github.com/kubernetes/kubernetes/issues/83913, but as you pointed out in https://github.com/kubernetes/kubernetes/issues/83913#issuecomment-548683208, that problem was already solved (differently, but still solved) in https://github.com/kubernetes/kubernetes/pull/84444, correct?

In that case we can simply close this PR unless someone sees some value in these enhancements.

okartau replied:
I am ok with closing.

cofyc replied:
HostExec is an alternative way to `ssh.SSH`. To avoid using kubectl, `*Framework` is still needed to access ExecWithOptions/ExecCommandInContainerWithFullOutput methods.

pohly replied:
@cofyc: so you are saying that https://github.com/kubernetes/kubernetes/pull/84444 does not remove the kubectl dependency completely and therefore this PR is still needed? That part is not clear to me.

But we have the source, so let's use that... :smile:

In current Kubernetes master, https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/testsuites/multivolume.go#L368 uses this chain of calls:
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L567
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L71
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L66
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/framework/util.go#L1109
- -> kubectl still needed

@okartau: I'm not sure why these new functions were necessary. Can you use the existing ExecWithOptions/ExecCommandInContainerWithFullOutput from https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/exec_util.go?

cofyc replied:
yes, this PR is still needed.

`(*Framework).ExecWithOptions/ExecCommandInContainerWithFullOutput` are used to exec command in an existing pod. However, `HostExec` is used to run commands on the host (by using `nsenter` to entering into host namespace). They are for different purposes.

cofyc commented on test/e2e/storage/utils/utils.go:
```diff
@@ -61,44 +62,47 @@ const (
+func PodExec(f *framework.Framework, pod *v1.Pod, shExec string) (string, error) {
+	clientSet := f.ClientSet
+	podClient := f.PodClient()
+	stdout, _, err := podClient.ExecCommandInContainerWithFullOutput(pod.Name, pod.Spec.Containers[0].Name, pod.Namespace, clientSet, "/bin/sh", "-c", shExec)
```
I think we can use `f.ExecCommandInContainerWithFullOutput` directly.

okartau replied:
indeed, seems it can, thanks for pointing that out. I changed to that call and patch set became smaller as two added-cloned functions are not needed any more.
Rebased also to HEAD of master and force-pushed.

pohly replied:
@cofyc: is it okay now?

cofyc replied:
yes, looks good to me (except one nit you pointed [here](https://github.com/kubernetes/kubernetes/pull/84042#discussion_r344926110))

pohly commented on test/e2e/storage/testsuites/volumes.go:
```diff
@@ -178,9 +178,9 @@ func (t *volumesTestSuite) defineTests(driver TestDriver, pattern testpatterns.T
 		// and we don't have reliable way to detect volumes are unmounted or
 		// not before starting the second pod.
-		volume.InjectContent(f.ClientSet, config, fsGroup, pattern.FsType, tests)
+		volume.InjectContent(f, f.ClientSet, config, fsGroup, pattern.FsType, tests)
```
Looks like we still have a redundant parameter here: `f.ClientSet` can be removed.

okartau replied:
removed clientset parameter, rebased and force-pushed.

okartau committed:
Removed clientset param from InjectContent

As one of previous changes added framework param to
InjectContent, it does not need separate param for clientset any more.

pohly commented:
/lgtm

pohly commented:
This is now entirely in code related to storage testing:
```
$ git diff 5dc87d2919818bbfd5c68eb1aed766b849a9062c..HEAD | diffstat
 common/volumes.go                   |    6 +-
 framework/volume/fixtures.go        |   24 +++++-----
 storage/flexvolume.go               |    8 +--
 storage/persistent_volumes-local.go |   40 ++++++++---------
 storage/testsuites/ephemeral.go     |    8 +--
 storage/testsuites/multivolume.go   |   24 +++++-----
 storage/testsuites/volume_io.go     |   22 ++++-----
 storage/testsuites/volumes.go       |    4 -
 storage/utils/utils.go              |   82 ++++++++++++++++++------------------
 storage/volumes.go                  |    2 
 upgrades/storage/volume_mode.go     |    6 +-
 11 files changed, 114 insertions(+), 112 deletions(-)
```

@msau42, perhaps you can approve?

pohly commented:
/assign @msau42 
for approval

msau42 commented:
/approve
